### PR TITLE
Laboratory Work #1, Yekaterina Golovleva - 20FPL3

### PR DESCRIPTION
**Pull request by esgolovleva**

### CODE REVIEW

File: ctlr_test.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import requests
+
+html_file = requests.get('https://www.m24.ru/news/').text
+with open('test.html', 'w', encoding='utf-8') as f:
+    f.write(html_file)
+
```
Comment (marina-kaz) [inline ctlr_test.py:6]: Nice, this should work in most cases. 
However, your source is rather tricky: try mimicking ordinary user requests by applying various headers and waiting for random time periods inbetween.

**Review by marina-kaz — CHANGES_REQUESTED** [review]
It is a start.

File: scrapper.py [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
 """
+import json
+import os
+
+import requests
+from bs4 import BeautifulSoup
+from random_user_agent.user_agent import UserAgent
+from random_user_agent.params import SoftwareName, OperatingSystem
+from constants import ASSETS_PATH
+from constants import CRAWLER_CONFIG_PATH
```
Comment (marina-kaz) [inline scrapper.py:12]: you can import it all at once
`from constants import ASSETS_PATH, CRAWLER_CONFIG_PATH`

Comment (marina-kaz) [reply scrapper.py:12]: also, it is better to import everything in alphabetical order

File: scrapper.py [hunk 3]
```diff
@@ -26,38 +36,123 @@ class Crawler:
...
+        operating_systems = [OperatingSystem.WINDOWS.value, OperatingSystem.LINUX.value]
+        useragent_random = UserAgent(software_names=software_names, operating_systems=operating_systems, limit=100)
+        headers = {'user-agent': useragent_random.get_random_user_agent(),
+                   'accept': '*/*', 'accept-language': 'ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7'}
+
+        for seed_url in self.seed_urls:
+            response = requests.get(seed_url, headers=headers, allow_redirects=True, timeout=5)
+            print(response.status_code)
+            with open('index1.html', 'w', encoding='utf-8') as file:
+                file.write(response.text)
```
Comment (marina-kaz) [inline scrapper.py:64]: This code is going to create a html file in the directory where scrapper.py is located. 
We do not need to save all the visited pages as html.
Refer to `docs/datset.md` to see where and how data must be stored.

File: scrapper.py [hunk 4]
```diff
@@ -26,38 +36,123 @@ class Crawler:
...
+        self.max_articles = max_articles
+        self.urls = []
 
     def _extract_url(self, article_bs):
-        pass
+        class_bs = article_bs.find('div', class_="b-materials-list b-list_infinity")
+        title_bs = class_bs.find_all('p', class_="b-materials-list__title b-materials-list__title_news")
+        for link in title_bs:
+            link = link.find('a')
+            self.urls.append('https://www.m24.ru' + link['href'])
```
Comment (marina-kaz) [inline scrapper.py:48]: it is better to use this method to solely extract urls from given soup
leave adding it to the attribute to `find_articles`
also, before adding a new url to the attribute it is necessary to check that such url has not been added before as many sources duplicate urls on their pages

File: scrapper.py [hunk 5]
```diff
@@ -26,38 +36,123 @@ class Crawler:
...
+        self.article = Article(article_url, article_id)
+
+    def parse(self):
+        response = requests.get(self.article_url)
+        # прописать обманку для сайта!!!
+        article_bs = BeautifulSoup(response.text, 'html.parser')
+
+        self._fill_article_with_text(article_bs)
+        self._fill_article_with_meta_information(article_bs)
+        self.article.save_raw()
```
Comment (marina-kaz) [inline scrapper.py:89]: it is not a responsibility of parser to save an article 
parser returns a filled instance of Article, and that is all

File: scrapper.py [hunk 6]
```diff
@@ -26,38 +36,123 @@ class Crawler:
...
+
+    def _fill_article_with_meta_information(self, article_bs):
 
 
 def prepare_environment(base_path):
     """
     Creates ASSETS_PATH folder if not created and removes existing folder
     """
-    pass
+    try:
```
Comment (marina-kaz) [inline scrapper.py:102]: this will work
however, is there a way to implement this without try/except tricks?

File: scrapper.py [hunk 7]
```diff
@@ -26,38 +36,123 @@ class Crawler:
...
+            raise IncorrectURLError
+        # возбуждает исключение, если такого seed_url
+        # нет в списке, который мы сделали в scrapper_config.json
+
+        if not isinstance(max_articles, int):
+            raise IncorrectNumberOfArticlesError
+        # если число статей, которое мы
+        # передаем в scrapper_config.json не является integer
+
+        if max_articles > 100:
```
Comment (marina-kaz) [inline scrapper.py:132]: I recommend taking a bigger number as 100 is a required number of articles to find and parse, so it is a bit odd that this code is expected to work for 100 items but is going to complain about 101

File: scrapper.py [hunk 8]
```diff
@@ -26,38 +36,123 @@ class Crawler:
...
 
 if __name__ == '__main__':
-    # YOUR CODE HERE
-    pass
+    seed_urls, max_articles = validate_config(crawler_path=CRAWLER_CONFIG_PATH)
+    prepare_environment(base_path=ASSETS_PATH)
+    crawler = Crawler(seed_urls=seed_urls, max_articles=max_articles)
+    crawler.find_articles()
+    print(crawler.urls)
+    # article.save_raw()
```
Comment (marina-kaz) [inline scrapper.py:156]: yep, that is exactly where saving should be :)

File: ctlr_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (marina-kaz) [reply ctlr_test.py:6]: We do not need this file anymore, do we?

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Keep going :)

File: ctlr_test.py (hunk 1 above)
Comment (esgolovleva) [reply ctlr_test.py:6]: done

[the remark this answers is not in the document]
Comment (esgolovleva) [reply ctlr_test.py]: I deleted this file

File: scrapper.py (hunk 2 above)
Comment (esgolovleva) [reply scrapper.py]: done

File: scrapper.py (hunk 3 above)
Comment (esgolovleva) [reply scrapper.py:63]: okay, I fixed that

File: scrapper.py (hunk 4 above)
Comment (esgolovleva) [reply scrapper.py]: done

File: scrapper.py (hunk 5 above)
Comment (esgolovleva) [reply scrapper.py:99]: okay, I try to follow SRP

File: scrapper.py (hunk 6 above)
Comment (esgolovleva) [reply scrapper.py]: thanks, i fixed that

File: scrapper.py (hunk 7 above)
Comment (esgolovleva) [reply scrapper.py]: okay, i'll take a bigger number of articles

File: scrapper.py (hunk 8 above)
Comment (esgolovleva) [reply scrapper.py]: nice!

File: scrapper.py [hunk 9]
```diff
@@ -1,6 +1,21 @@
 """
 Scrapper implementation
 """
+from datetime import datetime
+from pathlib import Path
+import random
+import re
+import shutil
+import time
+import json
```
Comment (marina-kaz) [inline scrapper.py:10]: it is a better practice to sort your imports alphabetically

File: scrapper.py (hunk 3 above)
Comment (marina-kaz) [reply scrapper.py:63]: as far as I can see, it is still there

File: scrapper.py [hunk 10]
```diff
@@ -26,38 +41,183 @@ class Crawler:
...
     def __init__(self, seed_urls, max_articles: int):
-        pass
+        self.seed_urls = seed_urls
+        self.max_articles = max_articles
+        self.urls = []
 
     def _extract_url(self, article_bs):
-        pass
+        link = article_bs.find('a')
+        return 'https://www.m24.ru' + link['href']
```
Comment (marina-kaz) [inline scrapper.py:50]: do you only extract one URL from each seed URL?

File: scrapper.py [hunk 11]
```diff
@@ -26,38 +41,183 @@ class Crawler:
...
+
+        head_title_bs = article_bs.find('title')
+        date_match = re.search(r'\d{2}\.\d{2}\.\d{4}', head_title_bs.text)
+        date_time_string = date_match.group(0)
+
+        time_bs = material_before_body.find('p', class_='b-material__date')
+        time_match = re.search(r'\d{2}:\d{2}', time_bs.text)
+        date_time_string = ' '.join([date_time_string, time_match.group(0)])
+
+        date_time = datetime.strptime(date_time_string, '%d.%m.%Y %H:%M')
```
Comment (marina-kaz) [inline scrapper.py:144]: You have a method for that in article.py 
Look it up

File: scrapper.py [hunk 12]
```diff
@@ -26,38 +41,183 @@ class Crawler:
...
+            raise IncorrectNumberOfArticlesError
+        # если число статей, которое мы
+        # передаем в scrapper_config.json не является integer
+
+        for seed_url in seed_urls:
+            if seed_url[0:8] != 'https://' and seed_url[0:7] != 'http://':
+                raise IncorrectURLError
+        # если протокол не соответствует стандратному паттерну
+
+        prepare_environment(ASSETS_PATH)  # пока не починили тесты
```
Comment (marina-kaz) [inline scrapper.py:196]: environment preparation is not a responsibility of config validator

**Review by marina-kaz — CHANGES_REQUESTED** [review]
You are almost there, time to speed up a little

File: scrapper.py (hunk 10 above)
Comment (esgolovleva) [reply scrapper.py]: i fixed that

File: scrapper.py (hunk 12 above)
Comment (esgolovleva) [reply scrapper.py]: i fixed that

File: scrapper.py (hunk 9 above)
Comment (marina-kaz) [reply scrapper.py]: Are you talking about the fact that pylint requires importing firstly standard packages, then external and then local? 
That is true, you should group them that way and import in this order.
However, within these blocks you are excepted to sort your imports alphabetically as this way your code is more readable

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Good job!! A little polishing and you are good to go

File: scrapper.py (hunk 3 above)
Comment (esgolovleva) [reply scrapper.py]: sorry, now I deleted these strings
all files stored here: 2021-1-level-ctlr\tmp\articles

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Congrats on successful Crawler submission!
Time to embark on Pipeline

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to work on Pipeline :)

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to speed up a little

**Review by marina-kaz — CHANGES_REQUESTED** [review]
No significant changes to review 
Deadline is approaching

File: pipeline.py [hunk 13]
```diff
@@ -72,31 +94,115 @@ class TextProcessingPipeline:
...
+            tokens_multi_tagged = []
+
+            for processed_token in processed_tokens:
+                tokens_cleaned.append(processed_token.get_cleaned())
+                tokens_single_tagged.append(processed_token.get_single_tagged())
+                tokens_multi_tagged.append(processed_token.get_multiple_tagged())
+
+            article.save_as(' '.join(tokens_cleaned), 'cleaned')
+            article.save_as(' '.join(tokens_single_tagged), 'single_tagged')
+            article.save_as(' '.join(tokens_multi_tagged), 'multiple_tagged')
```
Comment (marina-kaz) [inline pipeline.py]: as it is stated in the 5.3 section of pipeline documentation, you are required to pass `ArtifactType` attributes as a `kind` argument to the `save_as() method`, this way code is more secure as it is protected from almost untraceable errors caused by simple typos

File: pos_frequency_pipeline.py [hunk 14]
```diff
@@ -13,15 +18,45 @@ class EmptyFileError(Exception):
...
     def __init__(self, corpus_manager: CorpusManager):
-        pass
+        self.corpus_manager = corpus_manager
 
     def run(self):
-        pass
+        for article in self.corpus_manager.get_articles().values():
+            frequencies = {}
+            pattern = re.compile(r'<([A-Z]+)')
+            with open(article.get_file_path('single_tagged'), 'r') as file:
```
Comment (marina-kaz) [inline pos_frequency_pipeline.py]: the same thing goes here: instead of passing a plain string, use ArtifactType attribute

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Excellent job!

File: pipeline.py (hunk 13 above)
Comment (esgolovleva) [reply pipeline.py]: Thanks, I fixed that

File: pos_frequency_pipeline.py (hunk 14 above)
Comment (esgolovleva) [reply pos_frequency_pipeline.py]: I fixed that too

**Pull request closed without merging**